interactiveutil.jl tests

rofinn opened the pull request:
Added some tests for interactiveutil.jl
- Moved most of the edit function code into an editcmd function that returns the `run` or `spawn` function and the command string. This was to make it easier to test what editor is being selected.
- Currently, this only gets the interactiveutil coverage just above 50%. The remainder of the missing coverage is 1) OS dependent code 2) `gen_call_with_extracted_types(fcn, ex0)` (which isn't exported and only a small portion of the function is used) and 3) The `download`, `workspace` and `runtests` functions (which I'm not sure how best to unit test).

rofinn committed:
Added a couple basic tests to check that interactiveutils.jl doesn't have any glaring bugs and added interactiveutil tests to choosetests.

Added interactiveutil tests to choosetests

rofinn committed:
Split up edit(file::AbstractString, line::Integer) into2 separate functions (editcmd and edit) to make testing easier.

rofinn committed:
Added a couple tests for the clipboard and editcmd code and added editcmd to exports.jl

tkelman commented on base/exports.jl:
```diff
@@ -1076,6 +1076,7 @@ export
 # help and reflection
     apropos,
     current_module,
+    editcmd,
```
I don't think this is worth exporting

JeffBezanson replied:
In fact my preference would be to entirely disallow net increases in the list of Base exports.

Rory-Finnegan replied:
You're right, I should have just used `Base.editcmd` to access it.

tkelman commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,27 @@
+end
+
+# Not sure how detailed the versioninfo check needs to be
+versioninfo(true)
```
should redirect the output to avoid printing to screen during normal operation of the tests, wherever possible

tkelman commented:
There already is a [`workspace` test](https://github.com/JuliaLang/julia/blob/9390c71ace034e00e3c0a36837094bd89e465ac9/test/workspace.jl), but it's likely not showing up in the coverage statistics because it's running in a subprocess. I said this somewhere else recently, that test can probably check the status of `Base.JLOptions().code_coverage` to add that flag to the child julia process when the parent was called with coverage enabled.

tkelman commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,27 @@
+
+@unix_only begin
+    try
+        clipboard("Clipboard Test")
```
is it worth saving the existing contents first and restoring them in a `finally` block?

StefanKarpinski commented:
Seems like a good change minus the export of `editcmd`.

davidagold commented:
@tkelman it was here https://github.com/JuliaLang/julia/pull/12314#issuecomment-124996836

rofinn committed:
Several fixes to the interactiveutil test code.

1. removed editcmd from exports and switch to accessing via Base.editcmd
2. updated the version tests to using an IOBuffer rather than printing to STDOUT
3. moved the workspace.jl test into interactiveutil.jl to keep things organized and
   added a check to generate a coverage file for the workspace subprocess test if the tests are being run with coverage.

Rory-Finnegan commented:
Thanks for your suggestions everyone. I've added a few more commits to fix the mentioned issues (mostly in commit 325ff3c). I'll note that in commit 80df493 I removed a bunch of unused code in `gen_call_with_extracted_types`, so if something thinks I removed something really important please let me know.

tkelman commented on test/choosetests.jl:
```diff
@@ -15,7 +15,7 @@ Upon return, `tests` is a vector of fully-expanded test names, and
 function choosetests(choices = [])
     testnames = [
-        "linalg", "core", "keywordargs", "numbers", "printf",
+        "interactiveutil", "linalg", "core", "keywordargs", "numbers", "printf",
```
This shouldn't be first thing, sorry - linalg needs to stay first, and it's better for other time-consuming tests to be towards the start wherever possible. You can run just these tests by themselves via `make test-interactiveutil`

tkelman commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,75 @@
+@test contains(string(str), file)
+
+
+@unix_only begin
```
did this fail on appveyor if you tried to run it on windows?

tkelman commented on test/replcompletions.jl:
```diff
@@ -45,6 +46,7 @@ function temp_pkg_dir(fn::Function)
     fn()
   finally
     rm(tmpdir, recursive=true)
+    ENV = environment
```
This doesn't seem right - at the REPL this kind of thing gives a `WARNING: imported binding for ENV overwritten in module Main`. There's a `withenv` function that would probably be better for this.

tkelman commented on base/interactiveutil.jl:
```diff
@@ -229,46 +233,15 @@ typesof(args...) = Tuple{map(a->(isa(a,Type) ? Type{a} : typeof(a)), args)...}
 gen_call_with_extracted_types(fcn, ex0::Symbol) = Expr(:call, fcn, Meta.quot(ex0))
 function gen_call_with_extracted_types(fcn, ex0)
-    if isa(ex0, Expr) &&
-        any(a->(Meta.isexpr(a, :kw) || Meta.isexpr(a, :parameters)), ex0.args)
```
you sure this isn't needed? seems like it could be important for keyword-argument calls

tkelman commented on base/interactiveutil.jl:
```diff
@@ -229,46 +233,15 @@ typesof(args...) = Tuple{map(a->(isa(a,Type) ? Type{a} : typeof(a)), args)...}
-    exret = Expr(:call, :error, "expression is not a function call or symbol")
-    if !isa(ex, Expr)
-        # do nothing -> error
-    elseif ex.head == :call
```
this does look apparently redundant, but the `:body` and `:thunk` branches seem like they might still matter

tkelman commented:
Ah, I see in your commit message why you moved `interactiveutil` earlier. That doesn't seem like the right fix to me. What's the bug exactly with `versioninfo` when `JULIA_PKGDIR` doesn't exist? Seems like that could be fixed independently of the testing. And I'm a bit surprised environment variables that get set within test processes are leaking to other processes. If that's how that works, then it feels like it would've come up as an issue much sooner.

The changes to `gen_call_with_extracted_types` look a little too large to be doing as part of an otherwise-testing PR. Some of the code looks unused but not all, I wouldn't say your changes are obviously correct on inspection so would rather pull out most of that to a separate change.

Can anyone reproduce the linux-64 travis failure locally? Looks new to me:
```
	From worker 2:	     * serialize             in   3.13 seconds
exception on 2: 	From worker 4:	     * functors              in   0.51 seconds
	From worker 4:	     * enums                 in   1.36 seconds
ERROR: LoadError: test error in expression: contains(sprint((io->begin  # /tmp/julia/share/julia/test/misc.jl, line 6:
                info(io,"test")
            end)),"INFO:")
MethodError: `call` has no method matching call(::Base.AbstractIOBuffer{Array{UInt8,1}}, ::Base.AbstractIOBuffer{Array{UInt8,1}}, ::ASCIIString)
Closest candidates are:
  Union(!Matched::Any...)
  BoundsError(!Matched::Any...)
  TypeVar(!Matched::Any...)
  ...
 in anonymous at /tmp/julia/share/julia/test/misc.jl:6
 in sprint at ./iostream.jl:210
 in sprint at ./iostream.jl:214
 in anonymous at test.jl:92
 in do_test at test.jl:50
 in include_string at loading.jl:158
 in include_from_node1 at ./loading.jl:200
 in runtests at /tmp/julia/share/julia/test/testdefs.jl:196
 in anonymous at multi.jl:880
 in run_work_thunk at multi.jl:619
 in anonymous at task.jl:880
 in anonymous at /tmp/julia/share/julia/test/misc.jl:6
 in sprint at ./iostream.jl:210
 in sprint at ./iostream.jl:214
 in anonymous at test.jl:92
 in do_test at test.jl:50
 in include_string at loading.jl:158
 in include_from_node1 at ./loading.jl:200
 in runtests at /tmp/julia/share/julia/test/testdefs.jl:196
 in anonymous at multi.jl:880
 in run_work_thunk at multi.jl:619
 in anonymous at task.jl:880
while loading /tmp/julia/share/julia/test/misc.jl, in expression starting on line 6
	From worker 5:	     * ranges                in  69.39 seconds
	From worker 4:	     * cmdlineargs           in  52.92 seconds
	From worker 3:	     * subarray              in 335.90 seconds
ERROR: LoadError: LoadError: test error in expression: contains(sprint((io->begin  # /tmp/julia/share/julia/test/misc.jl, line 6:
                info(io,"test")
            end)),"INFO:")
MethodError: `call` has no method matching call(::Base.AbstractIOBuffer{Array{UInt8,1}}, ::Base.AbstractIOBuffer{Array{UInt8,1}}, ::ASCIIString)
Closest candidates are:
  Union(!Matched::Any...)
  BoundsError(!Matched::Any...)
  TypeVar(!Matched::Any...)
  ...
 in anonymous at task.jl:1397
while loading /tmp/julia/share/julia/test/misc.jl, in expression starting on line 6
while loading /tmp/julia/share/julia/test/runtests.jl, in expression starting on line 5
	From worker 2:	     * misc   
```

Rory-Finnegan commented:
@tkelman the error I was getting when calling versioninfo after JULIAPKGDIR gets changed was `ERROR: LoadError: Package metadata directory /var/folders/_6/25myjdtx2fxgjvznn19rp22m0000gn/T/lbeQm0ZF/v0.4/METADATA doesn't exist; run Pkg.init() to initialize it.` Changing the ordering of the tests and resetting the ENV seems to resolve it, even though that is a hacky solution. 

I'm fine with moving the change to `gen_call_with_extracted_types` to a different PR.

As for the new linux error I can't seem to reproduce it locally on linux and my personal travis builds outside this PR don't show it either. I'll try rebasing with the latest commits from master to see if that helps :/

Rory-Finnegan commented:
I can't seem to manually reproduce the JULIA_PKGDIR issue by running multiple processes that modify `ENV`, so I've removed my commits that try to fix it. However, you can see the error popped up in the new appveyor build, so maybe I should just make it a separate issue? Similarly, I've removed my `gen_call_with_extracted_types` changes and I'll make that a separate PR.

@tkelman removing the `unix_only` does fail on appveyor.

tkelman commented:
Cool, thanks for bearing with the review iteration. That `WARNING: imported binding for info overwritten in module Main` looks suspicious, wonder where that's coming from. It seems `versioninfo` could maybe be tweaked (in a separate issue) to not error out in this scenario.

`clipboard` failing on Windows surprises me a little, I thought we had an implementation there but it might be broken? Look at how many bugs adding tests is finding!

tkelman commented on test/interactiveutil.jl:
```diff
@@ -0,0 +1,63 @@
+end
+
+# Not sure how detailed the versioninfo check needs to be
+info = IOBuffer()
```
oh right, I guess this should use a different variable name, or a `let` block

iamed2 commented:
Perhaps this could be revisited with `@test_broken` for the broken test cases?
